File: app/src/main/java/ru/ulstu_team/ulstuschedule/ui/common/teacher/TeacherSchedulePresenter.java
package ru.ulstu_team.ulstuschedule.ui.common.teacher;

import java.util.List;

import javax.inject.Inject;

import io.realm.RealmQuery;
import ru.ulstu_team.ulstuschedule.data.DataManager;
import ru.ulstu_team.ulstuschedule.data.model.Lesson;
import ru.ulstu_team.ulstuschedule.data.remote.Schedule;
import ru.ulstu_team.ulstuschedule.data.remote.ScheduleRequest;
import ru.ulstu_team.ulstuschedule.ui.base.BasePresenter;

public class TeacherSchedulePresenter extends BasePresenter<TeacherScheduleMvpView> {

    private DataManager mDataManager;

    @Inject
    public TeacherSchedulePresenter(DataManager dataManager) {
        mDataManager = dataManager;
    }


    public void loadSchedule() {
        checkViewAttached();

        List<Lesson> lessons = getRealmQuery().findAll();
        if (lessons.isEmpty()) {
            //mDataManager.executeRequest(getRequest());
        } else {
            getMvpView().showSchedule(lessons);
        }
    }

    private RealmQuery<Lesson> getRealmQuery() {
        return getMRealm().where(Lesson.class).equalTo("TeacherId", mDataManager.getUserId());
    }

//    private ScheduleRequest getRequest() {
//        return new ScheduleRequest(Schedule.TEACHER_LESSONS,
//                mDataManager.getUserId(), Lesson.class, getRealmQuery(),
//                new ScheduleRequest.Callbacks() {
//                    @Override
//                    public void onSuccess() {
//                        List<Lesson> lessons = getRealmQuery().findAll();
//                        if (!lessons.isEmpty())
//                            getMvpView().showSchedule(lessons);
//                        else
//                            getMvpView().showEmptySchedule();
//                    }
//
//                    @Override
//                    public void onError(Exception e) {
//                        getMvpView().showError();
//                    }
//                }
//        );
//    }

    public void reload() {
        loadSchedule();
    }
}

File: app/src/main/java/ru/ulstu_team/ulstuschedule/ui/old/common/teacher/TeacherSchedulePresenter.kt
package ru.ulstu_team.ulstuschedule.ui.old.common.teacher

import javax.inject.Inject

import io.realm.RealmQuery
import ru.ulstu_team.ulstuschedule.data.DataManager
import ru.ulstu_team.ulstuschedule.data.model.Lesson
import ru.ulstu_team.ulstuschedule.data.remote.RequestCallbacks
import ru.ulstu_team.ulstuschedule.data.remote.Schedule
import ru.ulstu_team.ulstuschedule.data.remote.ScheduleRequest
import ru.ulstu_team.ulstuschedule.ui.base.BasePresenter

class TeacherSchedulePresenter
@Inject
constructor(private val mDataManager: DataManager) : BasePresenter<TeacherScheduleMvpView>() {


    fun loadScheduleForCurrentTeacher() {
        checkViewAttached()

        val lessons = mDataManager.getLessonsForCurrentTeacher()
        if (lessons.isNotEmpty())
            mvpView!!.showSchedule(lessons)
        else
            mDataManager.loadLessonsForCurrentTeacher(getCurrentTeacherRequestCallbacks())
    }

    fun loadScheduleForTeacher(teacherId: Int) {
        checkViewAttached()

        val lessons = mDataManager.getLessonsForTeacher(teacherId)
        if (lessons.isNotEmpty())
            mvpView!!.showSchedule(lessons)
        else
            mDataManager.loadLessonsForTeacher(teacherId, getTeacherRequestCallbacks(teacherId))
    }

    private fun getCurrentTeacherRequestCallbacks(): RequestCallbacks =
            object: RequestCallbacks {
                override fun onSuccess() {
                    val lessons = mDataManager.getLessonsForCurrentTeacher()
                    if (lessons.isNotEmpty())
                        mvpView!!.showSchedule(lessons)
                    else mvpView!!.showEmptySchedule()
                }

                override fun onError(e: Exception) {
                    mvpView!!.showError()
                }
            }

    private fun getTeacherRequestCallbacks(teacherId: Int): RequestCallbacks =
            object: RequestCallbacks {
                override fun onSuccess() {
                    val lessons = mDataManager.getLessonsForTeacher(teacherId)
                    if (lessons.isNotEmpty())
                        mvpView!!.showSchedule(lessons)
                    else
                        mvpView!!.showEmptySchedule()
                }

                override fun onError(e: Exception) {
                    mvpView!!.showError()
                }
            }
}
